---
permalink: /doc/samtools-view.1.html
layout: manpage
title: samtools-view(1) manual page
package: samtools-1.10
date: 6 December 2019
section_text: Bioinformatics tools
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools view &ndash; views and converts SAM/BAM/CRAM files
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
<b>view</b>
samtools <a href="samtools-view.1.html">view</a>
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[<em>region</em>...]
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
With no options or regions specified, prints all alignments in the specified
input alignment file (in SAM, BAM, or CRAM format) to standard output
in SAM format (with no header).
<p>
You may specify one or more space-separated region specifications after the
input filename to restrict output to only those alignments which overlap the
specified region(s). Use of region specifications requires a coordinate-sorted
and indexed input file (in BAM or CRAM format).
<p>
The
<b>-b</b>,
<b>-C</b>,
<b>-1</b>,
<b>-u</b>,
<b>-h</b>,
<b>-H</b>,
and
<b>-c</b>
options change the output format from the default of headerless SAM, and the
<b>-o</b>
and
<b>-U</b>
options set the output file name(s).
<p>
The
<b>-t</b>
and
<b>-T</b>
options provide additional reference data. One of these two options is required
when SAM input does not contain @SQ headers, and the
<b>-T</b>
option is required whenever writing CRAM output.
<p>
The
<b>-L</b>,
<b>-M</b>,
<b>-r</b>,
<b>-R</b>,
<b>-d</b>,
<b>-D</b>,
<b>-s</b>,
<b>-q</b>,
<b>-l</b>,
<b>-m</b>,
<b>-f</b>,
<b>-F</b>,
and
<b>-G</b>
options filter the alignments that will be included in the output to only those
alignments that match certain criteria.
<p>
The
<b>-x</b>
and
<b>-B</b>
options modify the data which is contained in each alignment.
<p>
The
<b>-X</b>
option can be used to allow user to specify customized index file location(s) if the data
folder does not contain any index file. See
<b>EXAMPLES</b>
section for sample of useage.
<p>
Finally, the
<b>-@</b>
option can be used to allocate additional threads to be used for compression, and the
<b>-?</b>
option requests a long help message.
<p>
<dl><dt><b>REGIONS:</b></dt><dd><p>
Regions can be specified as: RNAME[:STARTPOS[-ENDPOS]] and all position
coordinates are 1-based.
<p>
Important note: when multiple regions are given, some alignments may be output
multiple times if they overlap more than one of the specified regions.
<p>
Examples of region specifications:
<dl><dt><b>chr1</b></dt><dd><p>
Output all alignments mapped to the reference sequence named `chr1' (i.e. @SQ SN:chr1).
</dd><dt><b>chr2:1000000</b></dt><dd><p>
The region on chr2 beginning at base position 1,000,000 and ending at the
end of the chromosome.
</dd><dt><b>chr3:1000-2000</b></dt><dd><p>
The 1001bp region on chr3 beginning at base position 1,000 and ending at base
position 2,000 (including both end positions).
</dd><dt><b>'*'</b></dt><dd><p>
Output the unmapped reads at the end of the file.
(This does not include any unmapped reads placed on a reference sequence
alongside their mapped mates.)
</dd><dt><b>.</b></dt><dd><p>
Output all alignments.
(Mostly unnecessary as not specifying a region at all has the same effect.)
</dd></dl>
<p>
<p>
</dd></dl>
<p>
<h1 id="OPTIONS"><a href="#OPTIONS">OPTIONS</a></h1>
<dl><dt><b>-b</b></dt><dd><p>
Output in the BAM format.
</dd><dt><b>-C</b></dt><dd><p>
Output in the CRAM format (requires -T).
</dd><dt><b>-1</b></dt><dd><p>
Enable fast BAM compression (implies -b).
</dd><dt><b>-u</b></dt><dd><p>
Output uncompressed BAM. This option saves time spent on
compression/decompression and is thus preferred when the output is piped
to another samtools command.
</dd><dt><b>-h</b></dt><dd><p>
Include the header in the output.
</dd><dt><b>-H</b></dt><dd><p>
Output the header only.
</dd><dt><b>-c</b></dt><dd><p>
Instead of printing the alignments, only count them and print the
total number. All filter options, such as
<b>-f</b>,
<b>-F</b>,
and
<b>-q</b>,
are taken into account.
</dd><dt><b>-?</b></dt><dd><p>
Output long help and exit immediately.
</dd><dt><b>-o </b><em>FILE</em></dt><dd><p>
Output to
<em>FILE [stdout].</em>
</dd><dt><b>-U </b><em>FILE</em></dt><dd><p>
Write alignments that are
<em>not</em>
selected by the various filter options to
<em>FILE</em>.
When this option is used, all alignments (or all alignments intersecting the
<em>regions</em>
specified) are written to either the output file or this file, but never both.
</dd><dt><b>-t </b><em>FILE</em></dt><dd><p>
A tab-delimited
<em>FILE</em>.
Each line must contain the reference name in the first column and the length of
the reference in the second column, with one line for each distinct reference.
Any additional fields beyond the second column are ignored. This file also
defines the order of the reference sequences in sorting. If you run:
`samtools faidx &lt;ref.fa&gt;', the resulting index file
<em>&lt;ref.fa&gt;.fai</em>
can be used as this
<em>FILE</em>.
</dd><dt><b>-T </b><em>FILE</em></dt><dd><p>
A FASTA format reference
<em>FILE</em>,
optionally compressed by
<b>bgzip</b>
and ideally indexed by
<b>samtools</b>
<b>faidx</b>.
If an index is not present, one will be generated for you.
</dd><dt><b>-L </b><em>FILE</em></dt><dd><p>
Only output alignments overlapping the input BED
<em>FILE</em>
[null].
</dd><dt><b>-M </b></dt><dd><p>
Use the multi-region iterator on the union of a BED file and
command-line region arguments.  This avoids re-reading the same regions
of files so can sometimes be much faster.  Note this also removes
duplicate sequences.  Without this a sequence that overlaps multiple
regions specified on the command line will be reported multiple times.
The usage of a BED file is optional and its path has to be preceded by
<b>-L</b>
option.
</dd><dt><b>-r </b><em>STR</em></dt><dd><p>
Output alignments in read group
<em>STR</em>
[null].
Note that records with no
<b>RG</b>
tag will also be output when using this option.
This behaviour may change in a future release.
</dd><dt><b>-R </b><em>FILE</em></dt><dd><p>
Output alignments in read groups listed in
<em>FILE</em>
[null].
Note that records with no
<b>RG</b>
tag will also be output when using this option.
This behaviour may change in a future release.
</dd><dt><b>-d </b><em>STR:STR</em></dt><dd><p>
Only output alignments with tag
<em>STR</em>
and associated value
<em>STR</em>
[null].
</dd><dt><b>-D </b><em>STR:FILE</em></dt><dd><p>
Only output alignments with tag
<em>STR</em>
and associated values listed in
<em>FILE</em>
[null].
</dd><dt><b>-q </b><em>INT</em></dt><dd><p>
Skip alignments with MAPQ smaller than
<em>INT</em>
[0].
</dd><dt><b>-l </b><em>STR</em></dt><dd><p>
Only output alignments in library
<em>STR</em>
[null].
</dd><dt><b>-m </b><em>INT</em></dt><dd><p>
Only output alignments with number of CIGAR bases consuming query
sequence &ge;
<em>INT</em>
[0]
</dd><dt><b>-f </b><em>INT</em></dt><dd><p>
Only output alignments with all bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-F </b><em>INT</em></dt><dd><p>
Do not output alignments with any bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-G </b><em>INT</em></dt><dd><p>
Do not output alignments with all bits set in
<em>INT</em>
present in the FLAG field.  This is the opposite of <em>-f</em> such
that <em>-f12 -G12</em> is the same as no filtering at all.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-x </b><em>STR</em></dt><dd><p>
Read tag to exclude from output (repeatable) [null]
</dd><dt><b>-B</b></dt><dd><p>
Collapse the backward CIGAR operation.
</dd><dt><b>-s </b><em>FLOAT</em></dt><dd><p>
Output only a proportion of the input alignments.
This subsampling acts in the same way on all of the alignment records in
the same template or read pair, so it never keeps a read but not its mate.
<p>
The integer and fractional parts of the
<b>-s </b><em>INT</em><b>.</b><em>FRAC</em>
option are used separately: the part after the
decimal point sets the fraction of templates/pairs to be kept,
while the integer part is used as a seed that influences
<em>which</em>
subset of reads is kept.
<p>
When subsampling data that has previously been subsampled, be sure to use
a different seed value from those used previously; otherwise more reads
will be retained than expected.
</dd><dt><b>-@ </b><em>INT</em></dt><dd><p>
Number of BAM compression threads to use in addition to main thread [0].
</dd><dt><b>-S</b></dt><dd><p>
Ignored for compatibility with previous samtools versions.
Previously this option was required if input was in SAM format, but now the
correct format is automatically detected by examining the first few characters
of input.
</dd><dt><b>-X</b></dt><dd><p>
Include customized index file as a part of arugments. See
<b>EXAMPLES</b>
section for sample of useage.
</dd><dt><b>--no-PG</b></dt><dd><p>
Do not add a @PG line to the header of the output file.
<p>
</dd></dl>
<p>
<h1 id="EXAMPLES"><a href="#EXAMPLES">EXAMPLES</a></h1>
<ul><li>Import SAM to BAM when
<b>@SQ</b>
lines are present in the header:
<pre>
samtools view -bS aln.sam &gt; aln.bam
</pre>
If
<b>@SQ</b>
lines are absent:
<pre>
samtools faidx ref.fa
samtools view -bt ref.fa.fai aln.sam &gt; aln.bam
</pre>
where
<em>ref.fa.fai</em>
is generated automatically by the
<b>faidx</b>
command.
<p>
</li>
<li>Convert a BAM file to a CRAM file using a local reference sequence.
<pre>
samtools view -C -T ref.fa aln.bam &gt; aln.cram
</pre>
<p>
</li>
<li>Convert a BAM file to a CRAM with NM and MD tags stored verbatim
rather than calculating on the fly during CRAM decode, so that mixed
data sets with MD/NM only on some records, or NM calculated using
different definitions of mismatch, can be decoded without change.  The
second command demonstrates how to decode such a file.  The request to
not decode MD here is turning off auto-generation of both MD and NM;
it will still emit the MD/NM tags on records that had these stored
verbatim.
<pre>
samtools view -C --output-fmt-option store_md=1 --output-fmt-option store_nm=1 -o aln.cram aln.bam
samtools view --input-fmt-option decode_md=0 -o aln.new.bam aln.cram
</pre>
</li>
<li>An alternative way of achieving the above is listing multiple options
after the <b>--output-fmt</b> or <b>-O</b> option.  The commands below
are equivalent to the two above.
<pre>
samtools view -O cram,store_md=1,store_nm=1 -o aln.cram aln.bam
samtools view --input-fmt cram,decode_md=0 -o aln.new.bam aln.cram
</pre>
<p>
</li>
<li>Include customized index file as a part of arugments.
<pre>
samtools view [options] -X /data_folder/data.bam /index_folder/data.bai chrM:1-10
</pre>
<p>
</li>
<li>Output alignments in read group <b>grp2</b> (records with no <b>RG</b> tag will also be in the output).
<pre>
samtools view -r grp2 -o /data_folder/data.rg2.bam /data_folder/data.bam
</pre>
<p>
</li>
<li>Only keep reads with tag <b>BC</b> and were the barcode
matches the barcodes listed in the barcode file.
<pre>
samtools view -D BC:barcodes.txt -o /data_folder/data.barcodes.bam /data_folder/data.bam
</pre>
<p>
</li>
<li>Only keep reads with tag <b>RG</b> and read group <b>grp2</b>.
This does almost the same than <b>-r grp2</b> but will not keep records without the <b>RG</b> tag.
<pre>
samtools view -d RG:grp2 -o /data_folder/data.rg2_only.bam /data_folder/data.bam
</pre>
<p>
</ul>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Written by Heng Li from the Sanger Institute.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="samtools.1.html"><em>samtools</em></a> (1),
<a href="samtools-tview.1.html"><em>samtools-tview</em></a> (1),
<a href="sam.5.html"><em>sam</em></a> (5)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
